Removing logging config to avoid leaking environment secrets

MovieStoreGuy opened the pull request:
- The environment is loaded and printed out to stdout, this can not
happen in some environments and shouldn't be encouraged as good
practice.
- There is a few places the token is printed out in plain text, I had
removed such log lines

mpetazzoni commented on main.go:
```diff
@@ -903,11 +902,10 @@ func loadConfig(configFilePath string, logger log.Logger) (*config.GatewayConfig
-	var bb []byte
-	if bb, err = json.Marshal(loadedConfig); err == nil {
-		logger.Log(logkey.Config, string(bb), logkey.Env, strings.Join(os.Environ(), "-"), "config loaded")
+	if _, err = json.Marshal(loadedConfig); err != nil {
```
If we're not dumping the config anymore it might not be necessary to re-marshal it.

We should still offer a mechanism to preserve the previous behavior, for example through a command-line flag that asks the gateway to log its config, after calling `loadConfig()` in the `main()` method:

```go
loadedConfig, err := loadConfig(flags.configFileName, logger)
if &flags.showConfig {
  ... log the marshalled loadedConfig ...
}
```

So that one can run `gateway --show-config` or something like that.

MovieStoreGuy replied:
I wasn't paying too much attention what was happening within this section of code, and I assumed it was an additional validation step after loading all the variables from their various sources. 

I am really sceptical about adding any functionality that has the potential to leak secrets and would much rather the config to be manually inspected. However, knowing that some of the config is partially loaded from the environment, perhaps some info logs regarding that it has loaded info from the environment.

mdubbyap commented on main.go:
```diff
@@ -326,7 +327,6 @@ func (p *gateway) setupDebugServer(conf *config.GatewayConfig, logger log.Logger
 	})
 	p.debugServer.Mux.Handle("/debug/dims", &p.debugSink)
 
-	p.debugServer.Exp2.Exported["config"] = conf.Var()
```
This needs to stay.

MovieStoreGuy replied:
I'm curious to understand why this needs to stay?

Personal, I don't see the benefit since it doesn't reflect the health of the app and also contains the raw access token

mdubbyap commented on main.go:
```diff
@@ -903,11 +902,10 @@ func loadConfig(configFilePath string, logger log.Logger) (*config.GatewayConfig
-	var bb []byte
-	if bb, err = json.Marshal(loadedConfig); err == nil {
-		logger.Log(logkey.Config, string(bb), logkey.Env, strings.Join(os.Environ(), "-"), "config loaded")
+	if _, err = json.Marshal(loadedConfig); err != nil {
```
i agree with max, there's no benefit for re-marshalling it.

MovieStoreGuy commented:
Alrighty team, I finally had some time to get back this PR.

Taking into account all the feedback from everyone, I have added back the config var into the PR but ensures it removes secrets from being exposed. To further that, I have removed the unneeded marshal.

Though, I will leave someone else to add back the functionality to show the config with a command line arg.

mpetazzoni commented on config/config.go:
```diff
@@ -458,10 +458,20 @@ func (p *GatewayConfig) String() string {
+				settings.DefaultAuthToken = &redacted
+			}
+			if settings.AuthTokenEnvVar != nil {
+				settings.AuthTokenEnvVar = &redacted
```
Not sure I understand why this one needs to be redacted, since it doesn't contain a secret, but the name of the environment variable that does.

MovieStoreGuy replied:
Ooohh, right. Sorry, I saw it had token in the name and just redacted that.

mpetazzoni commented on config/config.go:
```diff
@@ -458,10 +458,20 @@ func (p *GatewayConfig) String() string {
 	return expvar.Func(func() interface{} {
-		return p
+		redacted := "redacted"
+		safeExport := p
```
Is that a deep clone?

MovieStoreGuy replied:
Yeah, I have a runable example here: https://play.golang.org/p/Mm7_aeO6Y5C

MovieStoreGuy commented:
Hi @mdubbyap ,

Just a gentle nudge to see if we can get these in?

mdubbyap commented:
I don't agree with this.  I *want* the token to be in the debug server if it's enabled. If you don't want the possibility to have the token observed this way, don't enable it.

MovieStoreGuy commented:
@mdubbyap what is the benefit of having the token in expvar config? 

I think goes beyond what you want and what is acceptable when running in an enterprise environment.

mdubbyap commented:
The debug server is there for that exact reason, debug, and in that sense i want to see everything that is there to be able to debug issues. 

In your hypothetical enterprise environment if you do not want the possibility that the token can be viewed this way, the you should not run the debug server. It is not on by default. I am firm on this.

mdubbyap commented:
I am submitting this as an alternative. https://github.com/signalfx/gateway/pull/239
